File: ios/runanywhere-sdks-latest/examples/flutter/RunAnywhereAI/lib/features/voice/stt_view_model.dart
import 'dart:async';

import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart' show IconData, Icons;
import 'package:runanywhere/runanywhere.dart' as sdk;

import 'package:runanywhere_ai/features/models/model_types.dart';
import 'package:runanywhere_ai/features/voice/voice_component_view_model_base.dart';

/// STTMode enumeration (matching iOS STTMode in STTViewModel.swift).
enum STTMode {
  batch,
  live,
  hybrid;

  String get displayName {
    switch (this) {
      case STTMode.batch:
        return 'Batch';
      case STTMode.live:
        return 'Live';
      case STTMode.hybrid:
        return 'Hybrid';
    }
  }

  String get description {
    switch (this) {
      case STTMode.batch:
        return 'Record first, then transcribe all at once';
      case STTMode.live:
        return 'Stream with live partial results';
      case STTMode.hybrid:
        return 'On-device first with cloud fallback';
    }
  }

  IconData get icon {
    switch (this) {
      case STTMode.batch:
        return Icons.mic;
      case STTMode.live:
        return Icons.stream;
      case STTMode.hybrid:
        return Icons.cloud;
    }
  }
}

/// ViewModel for the Speech-to-Text view (mirrors iOS `STTViewModel.swift`).
///
/// Owns recording, batch/live/hybrid transcription, hybrid cloud routing
/// config + router cache, and model selection state. The view only renders
/// state and forwards user intent.
class STTViewModel extends VoiceComponentViewModelBase {
  STTViewModel();

  // SDK-owned microphone capture (mirrors Swift AudioCaptureManager).
  final sdk.AudioCaptureManager _capture = sdk.AudioCaptureManager();
  StreamSubscription<double>? _audioLevelSubscription;

  // Live mode: mic chunks are fed straight into the SDK's streaming
  // transcription session (RunAnywhere.transcribeStream), which owns
  // endpointing/segmentation natively. No app-side silence detection.
  StreamSubscription<sdk.STTPartialResult>? _liveSubscription;
  String _committedTranscription = '';

  // Hybrid mode: router cached on the config tuple (mirrors iOS
  // `ensureHybridRouter`).
  sdk.HybridSttRouter? _hybridRouter;
  String? _hybridPairKey;

  // --- Component identity -----------------------------------------------------

  @override
  sdk.SDKComponent get component => sdk.SDKComponent.SDK_COMPONENT_STT;

  @override
  sdk.EventCategory get eventCategory => sdk.EventCategory.EVENT_CATEGORY_STT;

  @override
  ModelCategory get modelCategory =>
      ModelCategory.MODEL_CATEGORY_SPEECH_RECOGNITION;

  // --- UI state ----------------------------------------------------------------

  /// Committed transcription text shown in the main text area.
  String transcription = '';

  /// In-progress (non-final) live partial preview.
  String partialText = '';

  bool isRecording = false;
  bool isProcessing = false;
  bool isTranscribing = false;
  double audioLevel = 0.0;

  /// Sherpa-ONNX supports streaming transcription via its zipformer/RNN-T
  /// runtime. Allow Live mode for Sherpa-/ONNX-backed STT models.
  bool supportsLiveMode = true;

  // --- Hybrid cloud configuration (mirrors iOS STTViewModel) -------------------

  String cloudProviderId = 'flutter-demo-cloud-stt';
  String cloudProvider = sdk.CloudBackend.defaultProvider;
  String cloudModel = 'saarika:v2.5';
  String cloudApiKey = '';
  String cloudLanguageCode = 'en-IN';

  bool _hybridPreferOnline = false;
  bool get hybridPreferOnline => _hybridPreferOnline;
  set hybridPreferOnline(bool value) {
    _hybridPreferOnline = value;
    notify();
  }

  bool _hybridRequireNetwork = true;
  bool get hybridRequireNetwork => _hybridRequireNetwork;
  set hybridRequireNetwork(bool value) {
    _hybridRequireNetwork = value;
    notify();
  }

  /// Minimum battery percent for the cloud candidate (hard filter). Not
  /// surfaced in the UI (iOS parity) but always applied to the policy.
  double hybridMinBattery = 20;

  double _hybridConfidenceThreshold = sdk.kHybridSttConfidenceThreshold;
  double get hybridConfidenceThreshold => _hybridConfidenceThreshold;
  set hybridConfidenceThreshold(double value) {
    _hybridConfidenceThreshold = value;
    notify();
  }

  /// Routing decision of the last hybrid transcription, for display.
  sdk.HybridRoutedMetadata? hybridRouting;

  // --- Mode selection -----------------------------------------------------------

  STTMode _selectedMode = STTMode.batch;
  STTMode get selectedMode => _selectedMode;
  set selectedMode(STTMode mode) {
    if (_selectedMode == mode) return;
    final oldMode = _selectedMode;
    _selectedMode = mode;
    // Stop any active recording/transcription when mode changes (iOS parity).
    if (isRecording) {
      debugPrint('Mode changed from ${oldMode.name} to ${mode.name} - '
          'stopping active recording');
      unawaited(stopRecording());
    }
    if (oldMode == STTMode.live) {
      _stopLiveTranscription();
    }
    notify();
  }

  // --- Initialization -------------------------------------------------------------

  /// Initialize the ViewModel — idempotent. Microphone permission is requested
  /// by the view (the app's PermissionService needs a BuildContext).
  Future<void> initialize() async {
    if (!beginInitialization()) return;

    debugPrint('Initializing STT view model');
    subscribeToSDKEvents();
    await checkInitialModelState();
  }

  /// Load model from the model selection sheet.
  Future<void> loadModelFromSelection(ModelInfo model) async {
    isProcessing = true;
    notify();
    await loadModel(model);
    isProcessing = false;
    notify();
  }

  @override
  Future<void> performLoad(ModelInfo model) =>
      sdk.RunAnywhere.stt.load(model.id);

  /// STT resolves the display name from the model catalog when available,
  /// falling back to the model's own metadata (mirrors iOS).
  @override
  void applyLoadedModel(ModelInfo model) {
    applyLoadedModelFromCatalog(model);
    supportsLiveMode = selectedFramework ==
            LLMFramework.INFERENCE_FRAMEWORK_UNKNOWN ||
        selectedFramework == LLMFramework.INFERENCE_FRAMEWORK_ONNX;
    notify();
  }

  // --- Recording -------------------------------------------------------------------

  /// Toggle recording state (start/stop).
  Future<void> toggleRecording() async {
    if (isRecording) {
      await stopRecording();
    } else {
      await startRecording();
    }
  }

  Future<void> startRecording() async {
    debugPrint('Starting recording in ${_selectedMode.name} mode');
    errorMessage = null;
    hybridRouting = null;
    transcription = '';
    partialText = '';
    _committedTranscription = '';

    if (selectedModelId == null) {
      errorMessage = 'No STT model loaded';
      notify();
      return;
    }

    // iOS parity: hybrid mode requires a configured cloud API key.
    if (_selectedMode == STTMode.hybrid && !isHybridCloudConfigValid) {
      errorMessage = 'Enter a cloud STT API key before using Hybrid mode';
      notify();
      return;
    }

    if (_selectedMode == STTMode.live) {
      await _startLiveStreaming();
      return;
    }

    // Batch/hybrid: record to a buffer, transcribe once on stop.
    final recordingPath = await _capture.startRecordingToBuffer(
      sampleRate: 16000,
      numChannels: 1,
    );
    if (recordingPath == null) {
      errorMessage = 'Failed to start recording';
      notify();
      return;
    }

    _subscribeToAudioLevels();
    isRecording = true;
    notify();
    debugPrint('Recording started in ${_selectedMode.name} mode');
  }

  Future<void> stopRecording() async {
    debugPrint('Stopping recording');

    await _audioLevelSubscription?.cancel();
    _audioLevelSubscription = null;
    isRecording = false;
    audioLevel = 0.0;

    if (_selectedMode == STTMode.live) {
      // Closing the recorder ends the chunk stream, which lets the native
      // session flush its final result; the live subscription ends with it.
      isTranscribing = true;
      notify();
      await _capture.stopRecording();
      return;
    }
    notify();

    final audioData = await _capture.stopRecording();
    if (audioData == null || audioData.isEmpty) {
      errorMessage = 'No audio data recorded';
      notify();
      return;
    }

    if (_selectedMode == STTMode.hybrid) {
      await _performHybridTranscription(audioData);
    } else {
      await _performBatchTranscription(audioData);
    }
  }

  // --- Transcription ------------------------------------------------------------------

  /// Perform batch transcription on collected audio.
  Future<void> _performBatchTranscription(Uint8List audioData) async {
    debugPrint('Starting batch transcription of ${audioData.length} bytes');
    isTranscribing = true;
    errorMessage = null;
    notify();

    try {
      if (!sdk.RunAnywhere.stt.isLoaded) {
        throw StateError(
            'STT component not loaded. Please load an STT model first.');
      }
      final result = await sdk.RunAnywhere.stt.transcribe(audioData);
      transcription = result.text;
      debugPrint('Batch transcription complete: ${result.text.length} chars');
    } catch (e) {
      debugPrint('Batch transcription failed: $e');
      errorMessage = 'Transcription failed: $e';
    }

    isTranscribing = false;
    notify();
  }

  /// Perform one request through the SDK hybrid STT router
  /// (mirrors iOS `performHybridTranscription`).
  Future<void> _performHybridTranscription(Uint8List audioBytes) async {
    final offlineModelId = selectedModelId;
    if (offlineModelId == null) {
      errorMessage = 'No STT model loaded';
      notify();
      return;
    }

    debugPrint('Starting hybrid transcription of ${audioBytes.length} bytes');
    isTranscribing = true;
    errorMessage = null;
    hybridRouting = null;
    notify();

    try {
      final onlineModelId = _registerCloudProvider();
      final router = _ensureHybridRouter(
        offlineModelId: offlineModelId,
        onlineModelId: onlineModelId,
      );
      // Buffered recordings are WAV-encoded at 16 kHz (AudioCaptureManager).
      // audio_format wire values match rac_audio_format_enum_t (1 = WAV).
      final response = router.transcribe(
        audioBytes,
        options: sdk.HybridSttTranscribeOptions(
          sampleRate: 16000,
          audioFormat: 1,
        ),
      );
      transcription = response.text;
      hybridRouting = response.routing;
      debugPrint('Hybrid transcription complete: ${response.text}');
    } catch (e) {
      debugPrint('Hybrid transcription failed: $e');
      errorMessage = 'Hybrid transcription failed: $e';
    }

    isTranscribing = false;
    notify();
  }

  /// Whether the cloud config is complete enough to attempt hybrid routing.
  bool get isHybridCloudConfigValid =>
      cloudProviderId.trim().isNotEmpty &&
      cloudProvider.trim().isNotEmpty &&
      cloudModel.trim().isNotEmpty &&
      cloudApiKey.trim().isNotEmpty;

  /// Register the cloud plugin + provider credentials and return the cloud
  /// registry id the router resolves the online side by.
  String _registerCloudProvider() {
    final id = cloudProviderId.trim();
    final provider = cloudProvider.trim();
    final model = cloudModel.trim();
    final apiKey = cloudApiKey.trim();
    final language = cloudLanguageCode.trim();

    if (id.isEmpty || provider.isEmpty || model.isEmpty || apiKey.isEmpty) {
      throw StateError(
          'Cloud provider id, provider, model, and API key are required');
    }

    sdk.RunAnywhere.hybrid.registerCloud();
    sdk.RunAnywhere.hybrid.cloud.register(
      id: id,
      provider: provider,
      model: model,
      apiKey: apiKey,
      languageCode: language.isEmpty ? null : language,
    );
    return id;
  }

  /// Reuse the cached router when the config tuple is unchanged; otherwise
  /// close the old one and build a fresh pairing (mirrors iOS).
  sdk.HybridSttRouter _ensureHybridRouter({
    required String offlineModelId,
    required String onlineModelId,
  }) {
    final key = [
      offlineModelId,
      onlineModelId,
      cloudProvider.trim(),
      '$_hybridPreferOnline',
      '$_hybridRequireNetwork',
      '${hybridMinBattery.round()}',
      '$_hybridConfidenceThreshold',
    ].join('|');

    final cached = _hybridRouter;
    if (cached != null && _hybridPairKey == key) {
      return cached;
    }

    _hybridRouter?.close();
    _hybridRouter = null;
    _hybridPairKey = null;

    final router = sdk.RunAnywhere.hybrid.createSttRouter();
    final filters = <sdk.HybridFilter>[
      if (_hybridRequireNetwork) const sdk.HybridFilter.network(),
      sdk.HybridFilter.battery(minPercent: hybridMinBattery.round()),
    ];
    router.setPair(
      offline: sdk.HybridModel.offlineSherpa(offlineModelId),
      online: sdk.HybridModel.onlineCloud(
        onlineModelId,
        provider: cloudProvider.trim(),
      ),
      policy: sdk.HybridRoutingPolicy(
        hardFilters: filters,
        cascade: sdk.HybridCascade.confidence(_hybridConfidenceThreshold),
        rank: _hybridPreferOnline
            ? sdk.HybridRankOrder.preferOnlineFirst
            : sdk.HybridRankOrder.preferLocalFirst,
      ),
    );
    _hybridRouter = router;
    _hybridPairKey = key;
    return router;
  }

  // --- Live streaming ---------------------------------------------------------------

  /// Live mode: feed mic chunks into the SDK's streaming transcription
  /// session. Non-final partials preview the current utterance; finals
  /// commit it as a line (mirrors iOS `handleLivePartial`).
  Future<void> _startLiveStreaming() async {
    final chunks = await _capture.startRecording(
      sampleRate: 16000,
      numChannels: 1,
    );
    if (chunks == null) {
      errorMessage = 'Failed to start streaming recording';
      notify();
      return;
    }

    _subscribeToAudioLevels();

    _liveSubscription = sdk.RunAnywhere.transcribeStream(chunks).listen(
      (partial) {
        final text = partial.text.trim();
        if (partial.isFinal) {
          // Stream errors surface as a terminal partial carrying the
          // failure text (see RunAnywhere.transcribeStream).
          if (text.startsWith('STT stream failed')) {
            errorMessage = text;
            notify();
            return;
          }
          if (text.isNotEmpty) {
            _committedTranscription = _committedTranscription.isEmpty
                ? text
                : '$_committedTranscription\n$text';
          }
          transcription = _committedTranscription;
          partialText = '';
        } else if (text.isNotEmpty) {
          partialText = text;
        }
        notify();
      },
      onError: (Object e) {
        errorMessage = 'Transcription failed: $e';
        notify();
      },
      onDone: () {
        _liveSubscription = null;
        isTranscribing = false;
        notify();
      },
    );

    isRecording = true;
    notify();
    debugPrint('Live streaming transcription started');
  }

  void _subscribeToAudioLevels() {
    _audioLevelSubscription = _capture.audioLevelStream?.listen((level) {
      audioLevel = level;
      notify();
    });
  }

  /// Stop live transcription resources (called when mode changes).
  void _stopLiveTranscription() {
    unawaited(_liveSubscription?.cancel());
    _liveSubscription = null;
  }

  // --- Transcript actions --------------------------------------------------------------

  void clearTranscription() {
    transcription = '';
    partialText = '';
    hybridRouting = null;
    notify();
  }

  // --- Cleanup ---------------------------------------------------------------------------

  @override
  void dispose() {
    _stopLiveTranscription();
    unawaited(_audioLevelSubscription?.cancel());
    _audioLevelSubscription = null;
    unawaited(_capture.dispose());
    _hybridRouter?.close();
    _hybridRouter = null;
    _hybridPairKey = null;
    super.dispose();
  }
}

File: ios/runanywhere-sdks-latest/examples/ios/RunAnywhereAI/RunAnywhereAI/Features/Voice/STTViewModel.swift
//
//  STTViewModel.swift
//  RunAnywhereAI
//
//  ViewModel for Speech-to-Text functionality
//  Handles all business logic for STT including recording, transcription, and model management
//

import Foundation
import RunAnywhere
import Combine
import os

/// ViewModel for Speech-to-Text view
/// Manages recording, transcription, model selection, and microphone permissions
@MainActor
class STTViewModel: VoiceComponentViewModelBase {
    private let audioCapture = AudioCaptureManager()

    // MARK: - Component Identity

    override var component: RASDKComponent { .stt }
    override var eventCategory: RAEventCategory { .stt }
    override var modelCategory: RAModelCategory { .speechRecognition }

    // MARK: - Published Properties (UI State)

    @Published var transcription: String = ""
    @Published var isRecording = false
    @Published var isProcessing = false
    @Published var isTranscribing = false
    @Published var audioLevel: Float = 0.0
    @Published var cloudProviderId = "runanywhere-cloud-stt"
    @Published var cloudProvider = Cloud.defaultProvider
    @Published var cloudModel = "saarika:v2.5"
    @Published var cloudAPIKey = ""
    @Published var cloudLanguageCode = "en-IN"
    @Published var hybridPreferOnline = false
    @Published var hybridRequireNetwork = true
    @Published var hybridMinBattery: Double = 20
    @Published var hybridConfidenceThreshold = Double(RAHybridSTTConfidenceThreshold)
    @Published var hybridRouting: HybridRoutedMetadata?
    @Published var selectedMode: STTMode = .batch {
        didSet {
            // Stop any active recording/transcription when mode changes
            if oldValue != selectedMode {
                Task { @MainActor [weak self] in
                    guard let self = self else { return }
                    if self.isRecording {
                        let msg = "Mode changed from \(oldValue.rawValue) to \(self.selectedMode.rawValue)"
                        self.logger.info("\(msg) - stopping active recording")
                        await self.stopRecording()
                    }
                    // Also clean up any lingering live transcription resources
                    if oldValue == .live {
                        await self.stopLiveTranscription()
                    }
                }
            }
        }
    }

    // MARK: - Private Properties

    /// Batch mode: accumulated audio transcribed once on stop.
    private var audioBuffer = Data()

    /// Live mode: mic chunks are fed straight into the SDK's streaming
    /// transcription session (`RunAnywhere.transcribeStream`), which owns
    /// endpointing/segmentation natively. No app-side silence detection.
    private var liveAudioContinuation: AsyncStream<Data>.Continuation?
    private var liveStreamTask: Task<Void, Never>?
    private var committedTranscription = ""
    private var hybridRouter: HybridSTTRouter?
    private var hybridPairKey: String?

    // MARK: - Initialization State (for idempotency)

    private var hasSubscribedToAudioLevel = false

    // MARK: - Initialization

    init() {
        super.init(loggerCategory: "STT")
        logger.debug("STTViewModel initialized")
    }

    // MARK: - Public Methods

    /// Initialize the ViewModel - request permissions and setup subscriptions
    /// This method is idempotent - calling it multiple times is safe
    func initialize() async {
        guard beginInitialization() else { return }

        logger.info("Initializing STT view model")

        // Request microphone permission
        let hasPermission = await requestMicrophonePermission()
        if !hasPermission {
            errorMessage = "Microphone permission denied"
            logger.error("Microphone permission denied")
            return
        }

        // Subscribe to audio level updates (for batch mode)
        subscribeToAudioLevelUpdates()

        // Subscribe to SDK events for STT model state
        subscribeToSDKEvents()

        // Check initial STT model state
        await checkInitialModelState()
    }

    /// Load model from ModelSelectionSheet selection
    func loadModelFromSelection(_ model: RAModelInfo) async {
        isProcessing = true
        await loadModel(from: model)
        isProcessing = false
    }

    /// Toggle recording state (start/stop)
    func toggleRecording() async {
        if isRecording {
            await stopRecording()
        } else {
            await startRecording()
        }
    }

    // MARK: - Private Methods - Permissions

    private func requestMicrophonePermission() async -> Bool {
        await audioCapture.requestPermission()
    }

    // MARK: - Private Methods - Subscriptions

    private func subscribeToAudioLevelUpdates() {
        guard !hasSubscribedToAudioLevel else {
            logger.debug("Already subscribed to audio level updates, skipping")
            return
        }
        hasSubscribedToAudioLevel = true

        audioCapture.$audioLevel
            .receive(on: DispatchQueue.main)
            .sink { [weak self] level in
                // Defer state modifications to avoid "Publishing changes within view updates" warning
                Task { @MainActor in
                    self?.audioLevel = level
                }
            }
            .store(in: &cancellables)
    }

    /// STT resolves the display name from the model catalog when available,
    /// falling back to the id-derived name.
    override func applyLoadedModel(_ model: RAModelInfo) {
        selectedModelId = model.id
        if let matchingModel = ModelListViewModel.shared.availableModels.first(where: { $0.id == model.id }) {
            selectedModelName = matchingModel.name
            selectedFramework = matchingModel.framework
        } else {
            selectedModelName = model.id.modelNameFromID()
            selectedFramework = model.framework
        }
    }

    // MARK: - Private Methods - Recording

    private func startRecording() async {
        logger.info("Starting recording in \(self.selectedMode.rawValue) mode")
        errorMessage = nil
        hybridRouting = nil
        audioBuffer = Data()
        transcription = ""
        committedTranscription = ""

        guard selectedModelId != nil else {
            errorMessage = "No STT model loaded"
            return
        }

        if selectedMode == .hybrid && !isHybridCloudConfigValid {
            errorMessage = "Enter a cloud STT API key before using Hybrid mode"
            return
        }

        if selectedMode == .live {
            startLiveTranscription()
        }

        do {
            // Batch buffers locally; live feeds the SDK streaming session.
            try await AudioCapturePump.startRecording(with: audioCapture) { [weak self] audioData in
                guard let self else { return }
                if self.selectedMode == .live {
                    self.liveAudioContinuation?.yield(audioData)
                } else {
                    self.audioBuffer.append(audioData)
                }
            }

            isRecording = true
            logger.info("Recording started in \(self.selectedMode.rawValue) mode")
        } catch {
            logger.error("Failed to start recording: \(error.localizedDescription)")
            errorMessage = "Failed to start recording: \(error.localizedDescription)"
            await stopLiveTranscription()
        }
    }

    private func stopRecording() async {
        logger.info("Stopping recording")

        // Stop audio capture
        audioCapture.stopRecording()

        if selectedMode == .live {
            // Closing the audio stream lets the native session flush and emit
            // its final result; the consume task ends with the stream.
            liveAudioContinuation?.finish()
            liveAudioContinuation = nil
        } else if !audioBuffer.isEmpty {
            if selectedMode == .hybrid {
                await performHybridTranscription()
            } else {
                // Batch: transcribe everything we recorded.
                await performBatchTranscription()
            }
        }

        isRecording = false
        audioLevel = 0.0
    }

    // MARK: - Private Methods - Transcription

    /// Perform batch transcription on collected audio
    private func performBatchTranscription() async {
        guard !audioBuffer.isEmpty else {
            errorMessage = "No audio recorded"
            return
        }

        logger.info("Starting batch transcription of \(self.audioBuffer.count) bytes")
        isTranscribing = true
        transcription = ""

        do {
            let output = try await RunAnywhere.transcribe(audio: audioBuffer)
            transcription = output.text
            logger.info("Batch transcription complete: \(output.text)")
        } catch {
            logger.error("Batch transcription failed: \(error.localizedDescription)")
            errorMessage = "Transcription failed: \(error.localizedDescription)"
        }

        isTranscribing = false
    }

    /// Perform one request through the SDK hybrid STT router.
    private func performHybridTranscription() async {
        guard !audioBuffer.isEmpty else {
            errorMessage = "No audio recorded"
            return
        }
        guard let offlineModelId = selectedModelId else {
            errorMessage = "No STT model loaded"
            return
        }

        logger.info("Starting hybrid transcription of \(self.audioBuffer.count) bytes")
        isTranscribing = true
        transcription = ""
        hybridRouting = nil

        do {
            let onlineModelId = try registerCloudProvider()
            let router = try ensureHybridRouter(offlineModelId: offlineModelId, onlineModelId: onlineModelId)
            var options = HybridTranscribeOptions()
            options.sampleRate = 16_000
            options.audioFormat = CloudAudioFormat.wav.nativeValue

            let result = try router.transcribe(audioBuffer, options: options)
            transcription = result.text
            hybridRouting = result.routing
            logger.info("Hybrid transcription complete: \(result.text)")
        } catch {
            logger.error("Hybrid transcription failed: \(error.localizedDescription)")
            errorMessage = "Hybrid transcription failed: \(error.localizedDescription)"
        }

        isTranscribing = false
    }

    private var isHybridCloudConfigValid: Bool {
        !cloudProviderId.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty &&
            !cloudProvider.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty &&
            !cloudModel.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty &&
            !cloudAPIKey.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
    }

    private func registerCloudProvider() throws -> String {
        let id = cloudProviderId.trimmingCharacters(in: .whitespacesAndNewlines)
        let provider = cloudProvider.trimmingCharacters(in: .whitespacesAndNewlines)
        let model = cloudModel.trimmingCharacters(in: .whitespacesAndNewlines)
        let apiKey = cloudAPIKey.trimmingCharacters(in: .whitespacesAndNewlines)
        let language = cloudLanguageCode.trimmingCharacters(in: .whitespacesAndNewlines)

        guard !id.isEmpty, !provider.isEmpty, !model.isEmpty, !apiKey.isEmpty else {
            throw SDKException(
                code: .invalidArgument,
                message: "Cloud provider id, provider, model, and API key are required",
                category: .validation
            )
        }

        Cloud.register()
        Cloud.register(
            id: id,
            provider: provider,
            model: model,
            apiKey: apiKey,
            languageCode: language.isEmpty ? nil : language
        )
        return id
    }

    private func ensureHybridRouter(offlineModelId: String, onlineModelId: String) throws -> HybridSTTRouter {
        let key = [
            offlineModelId,
            onlineModelId,
            cloudProvider,
            String(hybridPreferOnline),
            String(hybridRequireNetwork),
            String(Int(hybridMinBattery)),
            String(hybridConfidenceThreshold),
        ].joined(separator: "|")

        if let router = hybridRouter, hybridPairKey == key {
            return router
        }

        hybridRouter?.close()
        let router = try HybridSTTRouter()
        var filters: [HybridFilter] = []
        if hybridRequireNetwork { filters.append(.network) }
        filters.append(.battery(minPercent: Int32(hybridMinBattery)))
        try router.setPair(
            offline: .offlineSherpa(offlineModelId),
            online: .onlineCloud(onlineModelId, provider: cloudProvider),
            policy: HybridRoutingPolicy(
                hardFilters: filters,
                cascade: .confidence(threshold: Float(hybridConfidenceThreshold)),
                rank: hybridPreferOnline ? .preferOnlineFirst : .preferLocalFirst
            )
        )
        hybridRouter = router
        hybridPairKey = key
        return router
    }

    /// Start the SDK streaming transcription session for live mode.
    ///
    /// Mic chunks are yielded into an `AsyncStream<Data>` consumed by
    /// `RunAnywhere.transcribeStream`; the native session owns segmentation
    /// and emits partial + final results.
    private func startLiveTranscription() {
        logger.info("Starting live streaming transcription")

        let (stream, continuation) = AsyncStream<Data>.makeStream()
        liveAudioContinuation = continuation

        liveStreamTask = Task { [weak self] in
            for await partial in RunAnywhere.transcribeStream(audio: stream) {
                guard let self, !Task.isCancelled else { break }
                self.handleLivePartial(partial)
            }
            self?.logger.info("Live transcription stream ended")
        }
    }

    /// Fold one streaming partial into the displayed transcription:
    /// non-final partials preview the current utterance, finals commit it.
    private func handleLivePartial(_ partial: RASTTPartialResult) {
        let text = partial.text.trimmingCharacters(in: .whitespacesAndNewlines)

        if partial.isFinal {
            // Stream errors surface as a terminal partial carrying the
            // failure text (see RunAnywhere.transcribeStream).
            if text.hasPrefix("STT stream failed") {
                errorMessage = text
                return
            }
            if !text.isEmpty {
                committedTranscription = committedTranscription.isEmpty
                    ? text
                    : committedTranscription + "\n" + text
            }
            transcription = committedTranscription
        } else if !text.isEmpty {
            transcription = committedTranscription.isEmpty
                ? text
                : committedTranscription + "\n" + text
        }
    }

    /// Stop live transcription (called when mode changes)
    private func stopLiveTranscription() async {
        logger.info("Stopping live transcription")
        liveAudioContinuation?.finish()
        liveAudioContinuation = nil
        liveStreamTask?.cancel()
        liveStreamTask = nil
    }

    // MARK: - Cleanup

    /// Clean up resources - call from view's onDisappear
    /// This replaces deinit cleanup to comply with Swift 6 concurrency
    func cleanup() {
        audioCapture.stopRecording()

        liveAudioContinuation?.finish()
        liveAudioContinuation = nil
        liveStreamTask?.cancel()
        liveStreamTask = nil
        hybridRouter?.close()
        hybridRouter = nil
        hybridPairKey = nil

        hasSubscribedToAudioLevel = false
        cleanupBase()
    }
}

// MARK: - Supporting Types

/// STT Mode for UI selection
enum STTMode: String {
    case batch
    case live
    case hybrid

    var icon: String {
        switch self {
        case .batch: return "square.stack.3d.up"
        case .live: return "waveform"
        case .hybrid: return "cloud"
        }
    }

    var description: String {
        switch self {
        case .batch: return "Record first, then transcribe"
        case .live: return "Stream with live partial results"
        case .hybrid: return "On-device first with cloud fallback"
        }
    }
}
